File: examples/C#/rtmama.cs
//
//  Custom routing Router to Mama (ROUTER to REQ)
//
//  While this example runs in a single process, that is just to make
//  it easier to start and stop the example. Each thread has its own
//  context and conceptually acts as a separate process.
//

//  Author:     Michael Compton, Tomas Roos
//  Email:      michael.compton@littleedge.co.uk, ptomasroos@gmail.com

using System;
using System.Collections.Generic;
using System.Text;
using ZMQ;
using System.Threading;

namespace ZMQGuide
{
    internal class Program
    {
        public static void Main(string[] args)
        {
            const int workersCount = 10;
            var workers = new List<Thread>(workersCount);
            
            using (var context = new Context(1))
            {
                using (Socket client = context.Socket(SocketType.ROUTER))
                {
                    client.Bind("tcp://*:5555");

                    for (int workerNumber = 0; workerNumber < workersCount; workerNumber++)
                    {
                        workers.Add(new Thread(WorkerTask));
                        workers[workerNumber].Start();
                    }

                    for (int taskNumber = 0; taskNumber < workersCount * 10; taskNumber++)
                    {
                        //  LRU worker is next waiting in queue
                        string address = client.Recv(Encoding.Unicode);
                        string empty = client.Recv(Encoding.Unicode);
                        string ready = client.Recv(Encoding.Unicode);

                        client.SendMore(address, Encoding.Unicode);
                        client.SendMore();
                        client.Send("This is the workload", Encoding.Unicode);
                    }

                    //  Now ask mamas to shut down and report their results
                    for (int taskNbr = 0; taskNbr < workersCount; taskNbr++)
                    {
                        string address = client.Recv(Encoding.Unicode);
                        string empty = client.Recv(Encoding.Unicode);
                        string ready = client.Recv(Encoding.Unicode);

                        client.SendMore(address, Encoding.Unicode);
                        client.SendMore();
                        client.Send("END", Encoding.Unicode);
                    }
                }
            }

            Console.ReadLine();
        }

        public static void WorkerTask()
        {
            var randomizer = new Random(DateTime.Now.Millisecond);

            using (var context = new Context(1))
            {
                using (Socket worker = context.Socket(SocketType.REQ))
                {
                    //  We use a string identity for ease here
                    ZHelpers.SetID(worker, Encoding.Unicode);
                    worker.Connect("tcp://localhost:5555");

                    int total = 0;

                    bool end = false;
                    while (!end)
                    {
                        //  Tell the router we're ready for work
                        worker.Send("Ready", Encoding.Unicode);

                        //  Get workload from router, until finished
                        string workload = worker.Recv(Encoding.Unicode);

                        if (workload.Equals("END"))
                        {
                            end = true;
                        }
                        else
                        {
                            total++;

                            Thread.Sleep(randomizer.Next(1, 1000)); //  Simulate 'work'
                        }
                    }

                    Console.WriteLine("ID ({0}) processed: {1} tasks", worker.IdentityToString(Encoding.Unicode), total);
                }
            }
        }
    }
}

File: examples/scala/rtmama.scala
/*
 *  Custom routing Router to Mama (ROUTER to REQ)
 *
 *  While this example runs in a single process, that is just to make
 *  it easier to start and stop the example. Each thread has its own
 *  context and conceptually acts as a separate process.
 *
 *
 * @author Giovanni Ruggiero
 * @email giovanni.ruggiero@gmail.com
 */

import org.zeromq.ZMQ
import ZHelpers._

object rtmama {

  class WorkerTask() extends Runnable {
    def run() {
			val rand = new java.util.Random(System.currentTimeMillis)
			val ctx = ZMQ.context(1)
			val worker = ctx.socket(ZMQ.REQ)
			//  We use a string identity for ease here
			setID(worker)
			// println(new String(worker.getIdentity))
			worker.connect("tcp://localhost:5555")
			var total = 0
			var workload = ""
      do {
				//  Tell the router we're ready for work
				worker.send("Ready".getBytes,0)
				workload = new String(worker.recv(0))
				Thread.sleep (rand.nextInt(1) * 1000)
				total += 1
				//  Get workload from router, until finished
			} while (!workload.equalsIgnoreCase("END"))
			printf("Processed: %d tasks\n", total)
		}
	}

  def main(args : Array[String]) {
		val NBR_WORKERS = 10
		val ctx = ZMQ.context(1)
		val client = ctx.socket(ZMQ.ROUTER)

		// Workaround to ckeck version >= 2.1
		assert(client.getType > -1)
		client.bind("tcp://*:5555")
		val workers = List.fill(NBR_WORKERS)(new Thread(new WorkerTask))
		workers foreach (_.start)

		for (i <- 1 to NBR_WORKERS * 10) {
			//  LRU worker is next waiting in queue
			val address = client.recv(0)
			val empty = client.recv(0)
			val ready = client.recv(0)

			client.send(address, ZMQ.SNDMORE)
			client.send("".getBytes, ZMQ.SNDMORE)
			client.send("This is the workload".getBytes,0)
		}

		//  Now ask mamas to shut down and report their results
		for (i <- 1 to NBR_WORKERS) {
			val address = client.recv(0)
			val empty = client.recv(0)
			val ready = client.recv(0)

			client.send(address, ZMQ.SNDMORE)
			client.send("".getBytes, ZMQ.SNDMORE)
			client.send("END".getBytes,0)
		}
	}
}
